File: Python/33-丑数.py
# -*- coding:utf-8 -*-
class Solution:
    def GetUglyNumber_Solution(self, index):
        # write code here
        if index < 7:
            return index
        id2, id3, id5 = 0, 0, 0
        res = [1]
        for _ in range(1, index):
            res.append(min(res[id2] * 2, res[id3] * 3, res[id5] * 5))
            if res[id2] * 2 == res[-1]: id2 += 1
            if res[id3] * 3 == res[-1]: id3 += 1
            if res[id5] * 5 == res[-1]: id5 += 1
        return res[index - 1]

File: C++/33-丑数.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution
{
public:
    int min(int a, int b, int c)
    {
        a = a < b ? a : b;
        return a < c ? a : c;
    }
    int GetUglyNumber_Solution(int index)
    {
        int id2 = 0, id3 = 0, id5 = 0;
        vector<int> res;
        res.push_back(1);
        for (size_t i = 1; i < index; i++)
        {
            res.push_back(min(res[id2] * 2, res[id3] * 3, res[id5] * 5));
            if (res[id2] * 2 == res.back())
                id2++;
            if (res[id3] * 3 == res.back())
                id3++;
            if (res[id5] * 5 == res.back())
                id5++;
        }
        return res[index - 1];
    }
};